src/server/pkg/cache/server/server.go

```diff
@@ -0,0 +1,114 @@
+package server
+
+import (
+	"hash/adler32"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache"
+	pb "github.com/golang/groupcache/groupcachepb"
+	"github.com/pachyderm/pachyderm/src/client/pkg/shard"
+	"github.com/pachyderm/pachyderm/src/server/pkg/cache/groupcachepb"
+	"go.pedge.io/proto/rpclog"
+	"golang.org/x/net/context"
+)
+
+// CacheServer serves groupcache requests over grpc.
+type CacheServer interface {
+	groupcachepb.GroupCacheServer
+	shard.Frontend
+	shard.Server
+}
+
+// NewCacheServer creates a new CacheServer.
+func NewCacheServer(router shard.Router, shards uint64) CacheServer {
+	server := &groupCacheServer{
+		Logger:      protorpclog.NewLogger("CacheServer"),
+		router:      router,
+		localShards: make(map[uint64]bool),
+		shards:      shards,
+	}
+	groupcache.RegisterPeerPicker(func() groupcache.PeerPicker { return server })
+	return server
+}
+
+type groupCacheServer struct {
+	protorpclog.Logger
+	router      shard.Router
+	localShards map[uint64]bool
+	mu          sync.Mutex
+	version     int64
+	shards      uint64
+}
+
+func (s *groupCacheServer) Get(
+	ctx context.Context,
+	request *groupcachepb.GetRequest,
+) (response *groupcachepb.GetResponse, retErr error) {
+	// response is too big to log
+	defer func(start time.Time) { s.Log(request, nil, retErr, time.Since(start)) }(time.Now())
```

Comment:
Do we want to follow the new paradigm where we also provide a log statement at the beginning of the function?

If I understand the get logic correctly, if its a cache miss the getter could try to read from obj storage and hang. So, it may be useful to instrument here as well w a log line at the beginning of the function.